src/EditorFeatures/Core/FindUsages/AggregateFindUsagesContext.cs

```diff
@@ -0,0 +1,125 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.FindUsages;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.FindUsages
+{
+    /// <summary>
+    /// Helper class that allows us to provide a unified Find-Usages progress over multiple
+    /// subroutines computing different types of usages.
+    /// </summary>
+    internal class FindUsagesContextAggregator
+    {
+        /// <summary>
+        /// Real context object we'll forward messages to.
+        /// </summary>
+        private readonly IFindUsagesContext _context;
+
+        /// <summary>
+        /// Underlying context objects that we'll receive messages from and forward to <see cref="_context"/>
+        /// </summary>
+        private readonly List<ForwardingFindUsagesContext> _underlyingContexts = new List<ForwardingFindUsagesContext>();
+
+        private readonly object _gate = new object();
+        private readonly Dictionary<ForwardingFindUsagesContext, (int current, int maxium)> _underlyingContextToProgress =
+            new Dictionary<ForwardingFindUsagesContext, (int current, int maxium)>();
+
+        public FindUsagesContextAggregator(IFindUsagesContext context)
+        {
+            _context = context;
+        }
+
+        public IFindUsagesContext CreateForwardingContext()
+        {
+            var result = new ForwardingFindUsagesContext(this);
+            lock (_gate)
+            {
+                _underlyingContexts.Add(result);
+            }
+            return result;
+        }
+
+        private Task ReportProgressAsync(ForwardingFindUsagesContext context, int current, int maximum)
+        {
+            int combinedCurrent = 0, combinedMaximum = 0;
+            lock (_gate)
+            {
+                _underlyingContextToProgress[context] = (current, maximum);
+
+                foreach (var (_, (singleCurrent, singleMaximum)) in _underlyingContextToProgress)
+                {
+                    combinedCurrent += singleCurrent;
+                    combinedMaximum += singleMaximum;
+                }
+            }
+
+            return _context.ReportProgressAsync(combinedCurrent, combinedMaximum);
```

Comment:
The threading implications of this is somewhat interesting: if two calls happen from two different producers there's no guarantee the "later" update will be the second call to the ReportProgressAsync; there's a window of time we might not report progress as far as we could.

Reply:
You're totally right.  I have fixed this up by serializing things.